eniius/mcstas.py

```diff
@@ -232,57 +232,68 @@ def getNXtype(cls, comp):
             return nxtype[0]
 
     @classmethod
-    def Slit(cls, **kw):
-        def to_float_or_string(value):
-            """Support string-named parameters; to allow replacement later"""
-            try:
-                return float(value)
-            except ValueError:
-                return value
+    def default_translation(cls, comp, **kw):
+        nxtype = cls.getNXtype(comp)
+        nx2mcs = NX2COMP_MAP.get(nxtype, ({}, {}))[1]
+        return getattr(nexus, nxtype)(**{nx: getattr(comp, mcs) for nx, mcs in nx2mcs.items()})
 
+    @classmethod
+    def Slit(cls, comp, **kw):
         def dif(a, b):
-            a = to_float_or_string(a)
-            b = to_float_or_string(b)
-            if isinstance(a, str) or isinstance(b, str):
-                return f"{a} - {b}"
-            return a - b
+            return f"{a} - {b}" if isinstance(a, str) or isinstance(b, str) else a - b
+
+        def neg_avg(a, b):
+            return f"-({a} + {b})/2" if isinstance(a, str) or isinstance(b, str) else (a + b) / 2
```

Comment:
Since the NeXus `NXslit` object is always symmetric about its reference point but the `McStas` `Slit` component may not be, if `a+b` is finite the `NXslit` needs to be translated to define the same opening.
Yesterday I thought the translation amount should be the negative value of the average 'edge' position -- but now I think I got that backwards.

If my current thinking is right, I'll replace `neg_avg` with
```
def avg(a, b):
    return f"({a} + {b})/2" if isinstance(a, str) or isinstance(b, str) else (a + b)/2
```

At the moment the extra translation is not added, in part because I was unsure if `NXtranslations` would accept a string-valued vector. The bigger question is, how to add even a numeric-valued extra translation?
- Should the component-named method optionally return an extra affine transformation?
- Should the method take the pre-defined transformation(s) group as input, and add the extra translation itself?

Reply:
Hmm... well a lot of the logic for dealing with transformations is not in the `McStasComp2NX` class where the method is, but is higher up in the `NXMcStas` class which calls `McStasComp2NX` - especially we would want to run `reduce_affine_transforms()` on the new concatenated list of previous transformations plus the new one from the component.

So it would probably make more sense to use the first option (the method return an optional extra transformation) but then to pass this back up to `NXMcStas`. E.g. we'd change [line 225](https://github.com/mducle/eniius/blob/8ea4888ebd061f51845ab3142e801e3b0ee26928/eniius/mcstas.py#L225) to:

```python
self.nxobj, self.extra_transform = method(mcstas_comp, **kwargs)
```

And then [line 588](https://github.com/mducle/eniius/blob/8ea4888ebd061f51845ab3142e801e3b0ee26928/eniius/mcstas.py#588) to something like:

```python
translated_component = McStasComp2NX(comp, order, self.NXtransformations(name), only_nx=only_nx, **mcpars)
if translated_component.extra_transform:   # Assuming the method returns None for extra_transforms if N/A
    self.affinelist[name] = reduce_affine_transforms(translated_component.extra_transform + self.affinelist[name])
    translated_component['transforms'] = self.NXtransformations(name)
return translated_component.nxobj
```

I think (!) that the extra transform should be applied last so the outer name for the `depends_on` is still ok... (not sure if I'm right though...)

Reply:
I've added a working implementation for this, which unfortunately requires the McStas instrument to contain the extra translation(s) in the component EXTEND block as `eniius_data`.
There's even a simple test to check that the whole thing works.

Getting it working in general without the extra `eniius_data` may not be feasible without parsing arbitrary C code to work-out the value of parameters.